File: ChatSupport/src/main/java/com/vantagecircle/chatapp/holder/ChatMViewHolder.java
package com.vantagecircle.chatapp.holder;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.FileProvider;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.Gravity;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.google.firebase.storage.StorageReference;
import com.vantagecircle.chatapp.R;
import com.vantagecircle.chatapp.core.FileHandler;
import com.vantagecircle.chatapp.core.interfaceC.FileInterface;
import com.vantagecircle.chatapp.core.model.FileModel;
import com.vantagecircle.chatapp.model.ChatM;
import com.vantagecircle.chatapp.services.SupportService;
import com.vantagecircle.chatapp.utils.ConfigUtils;
import com.vantagecircle.chatapp.utils.Constants;
import com.vantagecircle.chatapp.utils.DateUtils;
import com.vantagecircle.chatapp.utils.MainFileUtils;
import com.vantagecircle.chatapp.utils.UpdateKeyUtils;

import java.io.File;

/**
 * Created by bapidas on 27/07/17.
 */

public class ChatMViewHolder extends RecyclerView.ViewHolder {
    private final String TAG = ChatMViewHolder.class.getSimpleName();
    private TextView userName, messageText, dateTime;
    private ImageView statusImage, fileImage;
    private CardView lyt_thread;
    private ProgressBar progressBar;
    private LinearLayout lyt_parent;
    private Context context;

    public ChatMViewHolder(View itemView) {
        super(itemView);
        context = itemView.getContext();
        initViews(itemView);
    }

    private void initViews(View itemView) {
        userName = (TextView) itemView.findViewById(R.id.sender);
        messageText = (TextView) itemView.findViewById(R.id.text_content);
        dateTime = (TextView) itemView.findViewById(R.id.text_time);
        statusImage = (ImageView) itemView.findViewById(R.id.chat_status);
        fileImage = (ImageView) itemView.findViewById(R.id.image_status);
        progressBar = (ProgressBar) itemView.findViewById(R.id.progressBar);
        lyt_parent = (LinearLayout) itemView.findViewById(R.id.lyt_parent);
        lyt_thread = (CardView) itemView.findViewById(R.id.lyt_thread);
    }

    public void setDataToViews(ChatM chatM, boolean isChatContinue) {
        switch (chatM.getChatType()) {
            case Constants.IMAGE_CONTENT:
                if (chatM.getFileUrl() != null) {
                    if (MainFileUtils.isExternalLocal(chatM.getFileUrl())) {
                        downloadFile(chatM);
                    } else {
                        if (chatM.getSenderUid().equals(SupportService.id)) {
                            final File file = new File(Uri.parse(chatM.getFileUrl()).getPath());
                            if (file.exists()) {
                                ConfigUtils.loadPicasso(context, fileImage, chatM.getFileUrl());
                                fileImage.setOnClickListener(new View.OnClickListener() {
                                    @Override
                                    public void onClick(View v) {
                                        onFileClick(file);
                                    }
                                });
                                uploadFile(chatM);
                            } else {
                                fileImage.setImageResource(R.drawable.ic_insert_photo_black_24dp);
                                progressBar.setVisibility(View.GONE);
                            }
                        } else {
                            progressBar.setVisibility(View.VISIBLE);
                        }
                    }
                } else {
                    fileImage.setImageResource(R.drawable.ic_insert_photo_black_24dp);
                    progressBar.setVisibility(View.GONE);
                }
                break;
            default:
                messageText.setText(chatM.getMessageText());
                break;
        }

        userName.setText(chatM.getSenderName());
        dateTime.setText(DateUtils.getTimeAgo(chatM.getTimeStamp()));
        if (isChatContinue) {
            userName.setVisibility(View.GONE);
        } else {
            userName.setVisibility(View.VISIBLE);
        }

        //change row alignment on basis of user
        if (chatM.getSenderUid().equals(SupportService.id)) {
            userName.setTextColor(ContextCompat.getColor(context, R.color.colorOrange));
            lyt_parent.setPadding(100, 10, 15, 10);
            lyt_parent.setGravity(Gravity.END);
            statusImage.setVisibility(View.VISIBLE);
            lyt_thread.setCardBackgroundColor(ContextCompat.getColor(context, R.color.chat_background));
            if (chatM.isSentSuccessfully()) {
                if (chatM.isReadSuccessfully()) {
                    statusImage.setImageResource(R.drawable.tick_icon);
                } else {
                    statusImage.setImageResource(R.drawable.single_tick);
                }
            } else {
                statusImage.setImageResource(R.drawable.ic_msg_wait);
            }
        } else {
            userName.setTextColor(ContextCompat.getColor(context, R.color.colorAccent));
            statusImage.setVisibility(View.GONE);
            lyt_parent.setPadding(15, 10, 100, 10);
            lyt_parent.setGravity(Gravity.START);
            lyt_thread.setCardBackgroundColor(ContextCompat.getColor(context, R.color.white));
            if (!chatM.isReadSuccessfully()) {
                if (SupportService.getIsChatWindowActive()) {
                    UpdateKeyUtils.updateReadStatus(chatM.getChatRoom(), chatM.getTimeStamp());
                }
            }
        }
    }

    private void onFileClick(File file) {
        try {
            Intent intent = new Intent();
            intent.setAction(Intent.ACTION_VIEW);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                Uri contentUri = FileProvider.getUriForFile(context,
                        context.getPackageName() + ".fileProvider", file);
                intent.setDataAndType(contentUri, "*/*");
            } else {
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                intent.setDataAndType(Uri.fromFile(file), "*/*");
            }
            context.startActivity(intent);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void downloadFile(ChatM chatM) {
        FileHandler fileHandler = new FileHandler();
        final StorageReference storageReference = SupportService.getStorageInstance()
                .getReferenceFromUrl(chatM.getFileUrl());
        fileHandler.setStorageRef(storageReference);

        final String filepath;
        final String fileName;
        if (chatM.getChatType().contains("image")) {
            fileName = storageReference.getName();
        } else {
            fileName = "";
        }
        if (chatM.getSenderUid().equals(SupportService.id)) {
            filepath = MainFileUtils.getSentPath();
        } else {
            filepath = MainFileUtils.getReceivedPath();
        }

        try {
            final File file = MainFileUtils.isFilePresent(filepath, fileName);
            if (file != null) {
                ConfigUtils.loadPicasso(context, fileImage, Uri.fromFile(file).toString());
                fileImage.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        onFileClick(file);
                    }
                });
                progressBar.setVisibility(View.GONE);
            } else {
                File destination = new File(filepath, fileName);
                progressBar.setVisibility(View.VISIBLE);
                fileHandler.downloadFile(destination, new FileInterface() {
                    @Override
                    public void onProgress(FileModel fileModel) {
                        double progress = (100.0 * fileModel.getFileDownloadSnap().getBytesTransferred())
                                / fileModel.getFileDownloadSnap().getTotalByteCount();
                        Log.d(TAG, "Download is on progress  === " + progress + "%");
                    }

                    @Override
                    public void onPause(FileModel fileModel) {
                        Log.d(TAG, "Download is paused  === ");
                        progressBar.setVisibility(View.GONE);
                    }

                    @Override
                    public void onFail(FileModel fileModel) {
                        fileImage.setImageResource(R.drawable.ic_insert_photo_black_24dp);
                        progressBar.setVisibility(View.GONE);
                    }

                    @Override
                    public void onComplete(FileModel fileModel) {
                        try {
                            final File file = MainFileUtils.isFilePresent(filepath, fileName);
                            if (file != null) {
                                ConfigUtils.loadPicasso(context, fileImage, Uri.fromFile(file).toString());
                                fileImage.setOnClickListener(new View.OnClickListener() {
                                    @Override
                                    public void onClick(View v) {
                                        onFileClick(file);
                                    }
                                });
                            } else {
                                fileImage.setImageResource(R.drawable.ic_insert_photo_black_24dp);
                            }
                            progressBar.setVisibility(View.GONE);
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }
                });
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void uploadFile(final ChatM chatM) {
        FileHandler fileHandler = new FileHandler();
        Uri file = Uri.parse(chatM.getFileUrl());
        fileHandler.setStorageRef(SupportService.getChatImageReference()
                .child(MainFileUtils.getFileName(context, file)));

        progressBar.setVisibility(View.VISIBLE);
        fileHandler.uploadFile(file, chatM.getChatType(), new FileInterface() {
            @Override
            public void onProgress(FileModel fileModel) {
                double progress = (100.0 * fileModel.getUploadTaskSnap().getBytesTransferred())
                        / fileModel.getUploadTaskSnap().getTotalByteCount();
                Log.d(TAG, "Upload is on progress  === " + progress + "%");
            }

            @Override
            public void onPause(FileModel fileModel) {
                Log.d(TAG, "Upload is paused  === ");
                progressBar.setVisibility(View.GONE);
            }

            @Override
            public void onFail(FileModel fileModel) {
                Log.d(TAG, "Upload is failed  === " + fileModel.getException().getMessage());
                progressBar.setVisibility(View.GONE);
            }

            @Override
            public void onComplete(FileModel fileModel) {
                //get file download url
                String downloadUrl = fileModel.getUploadTaskSnap().getDownloadUrl().toString();
                //update file url to the server
                UpdateKeyUtils.updateFileUrl(chatM.getChatRoom(), chatM.getTimeStamp(), downloadUrl);

                //push notification after file update is complete
                /*chatM.setFileUrl(downloadUrl);
                SendNotification sendNotification = new SendNotification();
                sendNotification.prepareNotification(chatM);*/
            }
        });
    }
}

File: app/src/main/java/com/bapidas/chattingapp/ui/adapter/holder/ChatMViewHolder.kt
package com.bapidas.chattingapp.ui.adapter.holder

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.util.Log
import android.view.Gravity
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.recyclerview.widget.RecyclerView
import com.bapidas.chattingapp.ChatApplication
import com.bapidas.chattingapp.R
import com.bapidas.chattingapp.data.core.FileHandler
import com.bapidas.chattingapp.data.core.callbacks.FileInterface
import com.bapidas.chattingapp.data.core.model.FileModel
import com.bapidas.chattingapp.data.model.ChatM
import com.bapidas.chattingapp.notification.httpcall.SendNotification
import com.bapidas.chattingapp.utils.ConfigUtils.loadPicasso
import com.bapidas.chattingapp.utils.Constants
import com.bapidas.chattingapp.utils.DateUtils.getTimeAgo
import com.bapidas.chattingapp.utils.MainFileUtils.getFileName
import com.bapidas.chattingapp.utils.MainFileUtils.isExternalLocal
import com.bapidas.chattingapp.utils.MainFileUtils.isFilePresent
import com.bapidas.chattingapp.utils.MainFileUtils.receivedPath
import com.bapidas.chattingapp.utils.MainFileUtils.sentPath
import com.bapidas.chattingapp.utils.UpdateKeyUtils.updateFileUrl
import com.bapidas.chattingapp.utils.UpdateKeyUtils.updateReadStatus
import java.io.File

/**
 * Created by bapidas on 27/07/17.
 */
class ChatMViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
    private val TAG = ChatMViewHolder::class.java.simpleName
    private val context: Context = itemView.context

    private val lytParent: LinearLayout = itemView.findViewById<View>(R.id.lyt_parent) as LinearLayout
    private val lytThread: CardView = itemView.findViewById<View>(R.id.lyt_thread) as CardView
    private val userName: TextView = itemView.findViewById<View>(R.id.sender) as TextView
    private val messageText: TextView = itemView.findViewById<View>(R.id.text_content) as TextView
    private val dateTime: TextView = itemView.findViewById<View>(R.id.text_time) as TextView
    private val statusImage: ImageView = itemView.findViewById<View>(R.id.chat_status) as ImageView

    private var fileImage: ImageView? = null
    private var progressBar: ProgressBar? = null

    fun setDataToViews(chatM: ChatM, isChatContinue: Boolean) {
        when (chatM.chatType) {
            Constants.IMAGE_CONTENT ->  {
                progressBar = itemView.findViewById<View>(R.id.progressBar) as ProgressBar
                fileImage = itemView.findViewById<View>(R.id.image_status) as ImageView

                if (chatM.fileUrl.isNotEmpty()) {
                    if (isExternalLocal(chatM.fileUrl)) {
                        downloadFile(chatM)
                    } else {
                        if (chatM.senderUid == ChatApplication.applicationContext().id) {
                            val file = File(Uri.parse(chatM.fileUrl).path.orEmpty())
                            if (file.exists()) {
                                fileImage?.let { loadPicasso(context, it, chatM.fileUrl) }
                                fileImage?.setOnClickListener { onFileClick(file) }
                                uploadFile(chatM)
                            } else {
                                fileImage?.setImageResource(R.drawable.ic_insert_photo_black_24dp)
                                progressBar?.visibility = View.GONE
                            }
                        } else {
                            progressBar?.visibility = View.VISIBLE
                        }
                    }
                } else {
                    fileImage?.setImageResource(R.drawable.ic_insert_photo_black_24dp)
                    progressBar?.visibility = View.GONE
                }
            }
            else -> messageText.text = chatM.messageText
        }
        userName.text = chatM.senderName
        dateTime.text = getTimeAgo(chatM.timeStamp)
        if (isChatContinue) {
            userName.visibility = View.GONE
        } else {
            userName.visibility = View.VISIBLE
        }

        //change row alignment on basis of user
        if (chatM.senderUid == ChatApplication.applicationContext().id) {
            userName.setTextColor(ContextCompat.getColor(context, R.color.colorOrange))
            lytParent.setPadding(100, 10, 15, 10)
            lytParent.gravity = Gravity.END
            statusImage.visibility = View.VISIBLE
            lytThread.setCardBackgroundColor(ContextCompat.getColor(context, R.color.chat_background))
            if (chatM.isSentSuccessfully) {
                if (chatM.isReadSuccessfully) {
                    statusImage.setImageResource(R.drawable.tick_icon)
                } else {
                    statusImage.setImageResource(R.drawable.single_tick)
                }
            } else {
                statusImage.setImageResource(R.drawable.ic_msg_wait)
            }
        } else {
            userName.setTextColor(ContextCompat.getColor(context, R.color.colorAccent))
            statusImage.visibility = View.GONE
            lytParent.setPadding(15, 10, 100, 10)
            lytParent.gravity = Gravity.START
            lytThread.setCardBackgroundColor(ContextCompat.getColor(context, R.color.white))
            if (!chatM.isReadSuccessfully) {
                if (ChatApplication.applicationContext().isChatWindowActive) {
                    updateReadStatus(chatM.chatRoom, chatM.timeStamp)
                }
            }
        }
    }

    private fun onFileClick(file: File) {
        try {
            val intent = Intent()
            intent.action = Intent.ACTION_VIEW
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                intent.flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
                val contentUri = FileProvider.getUriForFile(context,
                        context.packageName + ".utils.chatFileProvider", file)
                intent.setDataAndType(contentUri, "*/*")
            } else {
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                intent.setDataAndType(Uri.fromFile(file), "*/*")
            }
            context.startActivity(intent)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun downloadFile(chatM: ChatM) {
        val fileHandler = FileHandler()
        val storageReference = ChatApplication.applicationContext().storageInstance
                .getReferenceFromUrl(chatM.fileUrl)
        fileHandler.storageReference = storageReference
        val fileName: String
        fileName = if (chatM.chatType.contains("image")) {
            storageReference.name
        } else {
            ""
        }
        val filepath: String = if (chatM.senderUid == ChatApplication.applicationContext().id) {
            sentPath
        } else {
            receivedPath
        }
        try {
            val file = isFilePresent(filepath, fileName)
            if (file != null) {
                fileImage?.let { loadPicasso(context, it, Uri.fromFile(file).toString()) }
                fileImage?.setOnClickListener { onFileClick(file) }
                progressBar?.visibility = View.GONE
            } else {
                val destination = File(filepath, fileName)
                progressBar?.visibility = View.VISIBLE
                fileHandler.downloadFile(destination, object : FileInterface {
                    override fun onProgress(fileModel: FileModel) {
                        val res = (fileModel.fileDownloadSnap?.bytesTransferred ?: 0) /
                                (fileModel.fileDownloadSnap?.totalByteCount ?: 0)
                        val progress = (100.0 * res)
                        Log.d(TAG, "Download is on progress  === $progress%")
                    }

                    override fun onPause(fileModel: FileModel) {
                        Log.d(TAG, "Download is paused  === ")
                        progressBar?.visibility = View.GONE
                    }

                    override fun onFail(fileModel: FileModel) {
                        fileImage?.setImageResource(R.drawable.ic_insert_photo_black_24dp)
                        progressBar?.visibility = View.GONE
                    }

                    override fun onComplete(fileModel: FileModel) {
                        try {
                            val newFile = isFilePresent(filepath, fileName)
                            if (newFile != null) {
                                fileImage?.let { loadPicasso(context, it, Uri.fromFile(newFile).toString()) }
                                fileImage?.setOnClickListener { onFileClick(newFile) }
                            } else {
                                fileImage?.setImageResource(R.drawable.ic_insert_photo_black_24dp)
                            }
                            progressBar?.visibility = View.GONE
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                    }
                })
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun uploadFile(chatM: ChatM) {
        val fileHandler = FileHandler()
        val file = Uri.parse(chatM.fileUrl)
        fileHandler.storageReference = ChatApplication.applicationContext()
                .chatImageReference.child(getFileName(context, file))
        progressBar?.visibility = View.VISIBLE
        fileHandler.uploadFile(file, chatM.chatType, object : FileInterface {
            override fun onProgress(fileModel: FileModel) {
                val res = (fileModel.fileDownloadSnap?.bytesTransferred ?: 0) /
                        (fileModel.fileDownloadSnap?.totalByteCount ?: 0)
                val progress = (100.0 * res)
                Log.d(TAG, "Upload is on progress  === $progress%")
            }

            override fun onPause(fileModel: FileModel) {
                Log.d(TAG, "Upload is paused  === ")
                progressBar?.visibility = View.GONE
            }

            override fun onFail(fileModel: FileModel) {
                Log.d(TAG, "Upload is failed  === " + fileModel.exception?.message)
                progressBar?.visibility = View.GONE
            }

            override fun onComplete(fileModel: FileModel) {
                //get file download url
                val downloadUrl = fileModel.uploadTaskSnap?.uploadSessionUri?.path
                if (!downloadUrl.isNullOrEmpty()) {
                    //update file url to the server
                    updateFileUrl(chatM.chatRoom, chatM.timeStamp, downloadUrl)

                    //push notification after file update is complete
                    chatM.fileUrl = downloadUrl
                    val sendNotification = SendNotification(context)
                    sendNotification.prepareNotification(chatM)
                }
            }
        })
    }
}